File: binding/include/libqstringlistmodel.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstringlistmodel.html
pub const qstringlistmodel = struct {

    /// New constructs a new QStringListModel object.
    pub fn New() QtC.QStringListModel {
        return qtc.QStringListModel_new();
    }


    /// New2 constructs a new QStringListModel object.
    pub fn New2(strings: []const u8) QtC.QStringListModel {
        return qtc.QStringListModel_new2(@ptrCast(strings));
    }


    /// New3 constructs a new QStringListModel object.
    pub fn New3(parent: ?*anyopaque) QtC.QStringListModel {
        return qtc.QStringListModel_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QStringListModel object.
    pub fn New4(strings: []const u8, parent: ?*anyopaque) QtC.QStringListModel {
        return qtc.QStringListModel_new4(@ptrCast(strings), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStringListModel_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstringlistmodel.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn RowCount(self: ?*anyopaque, parent: ?*anyopaque) i32 {
        return qtc.QStringListModel_RowCount(@ptrCast(self), @ptrCast(parent));
    }

    /// Allows for overriding the related default method
    pub fn OnRowCount(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) i32) void {
        qtc.QStringListModel_OnRowCount(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRowCount(self: ?*anyopaque, parent: ?*anyopaque) i32 {
        return qtc.QStringListModel_QBaseRowCount(@ptrCast(self), @ptrCast(parent));
    }

    pub fn Sibling(self: ?*anyopaque, row: i32, column: i32, idx: ?*anyopaque) QtC.QModelIndex {
        return qtc.QStringListModel_Sibling(@ptrCast(self), row, column, @ptrCast(idx));
    }

    /// Allows for overriding the related default method
    pub fn OnSibling(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32, ?*anyopaque) callconv(.c) QtC.QModelIndex) void {
        qtc.QStringListModel_OnSibling(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSibling(self: ?*anyopaque, row: i32, column: i32, idx: ?*anyopaque) QtC.QModelIndex {
        return qtc.QStringListModel_QBaseSibling(@ptrCast(self), row, column, @ptrCast(idx));
    }

    pub fn Data(self: ?*anyopaque, index: ?*anyopaque, role: i32) QtC.QVariant {
        return qtc.QStringListModel_Data(@ptrCast(self), @ptrCast(index), role);
    }

    /// Allows for overriding the related default method
    pub fn OnData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, i32) callconv(.c) QtC.QVariant) void {
        qtc.QStringListModel_OnData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseData(self: ?*anyopaque, index: ?*anyopaque, role: i32) QtC.QVariant {
        return qtc.QStringListModel_QBaseData(@ptrCast(self), @ptrCast(index), role);
    }

    pub fn SetData(self: ?*anyopaque, index: ?*anyopaque, value: ?*anyopaque, role: i32) bool {
        return qtc.QStringListModel_SetData(@ptrCast(self), @ptrCast(index), @ptrCast(value), role);
    }

    /// Allows for overriding the related default method
    pub fn OnSetData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QStringListModel_OnSetData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetData(self: ?*anyopaque, index: ?*anyopaque, value: ?*anyopaque, role: i32) bool {
        return qtc.QStringListModel_QBaseSetData(@ptrCast(self), @ptrCast(index), @ptrCast(value), role);
    }

    pub fn ClearItemData(self: ?*anyopaque, index: ?*anyopaque) bool {
        return qtc.QStringListModel_ClearItemData(@ptrCast(self), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnClearItemData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStringListModel_OnClearItemData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseClearItemData(self: ?*anyopaque, index: ?*anyopaque) bool {
        return qtc.QStringListModel_QBaseClearItemData(@ptrCast(self), @ptrCast(index));
    }

    pub fn Flags(self: ?*anyopaque, index: ?*anyopaque) i32 {
        return qtc.QStringListModel_Flags(@ptrCast(self), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnFlags(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) i32) void {
        qtc.QStringListModel_OnFlags(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseFlags(self: ?*anyopaque, index: ?*anyopaque) i32 {
        return qtc.QStringListModel_QBaseFlags(@ptrCast(self), @ptrCast(index));
    }

    pub fn InsertRows(self: ?*anyopaque, row: i32, count: i32, parent: ?*anyopaque) bool {
        return qtc.QStringListModel_InsertRows(@ptrCast(self), row, count, @ptrCast(parent));
    }

    /// Allows for overriding the related default method
    pub fn OnInsertRows(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStringListModel_OnInsertRows(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseInsertRows(self: ?*anyopaque, row: i32, count: i32, parent: ?*anyopaque) bool {
        return qtc.QStringListModel_QBaseInsertRows(@ptrCast(self), row, count, @ptrCast(parent));
    }

    pub fn RemoveRows(self: ?*anyopaque, row: i32, count: i32, parent: ?*anyopaque) bool {
        return qtc.QStringListModel_RemoveRows(@ptrCast(self), row, count, @ptrCast(parent));
    }

    /// Allows for overriding the related default method
    pub fn OnRemoveRows(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStringListModel_OnRemoveRows(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseRemoveRows(self: ?*anyopaque, row: i32, count: i32, parent: ?*anyopaque) bool {
        return qtc.QStringListModel_QBaseRemoveRows(@ptrCast(self), row, count, @ptrCast(parent));
    }

    pub fn MoveRows(self: ?*anyopaque, sourceParent: ?*anyopaque, sourceRow: i32, count: i32, destinationParent: ?*anyopaque, destinationChild: i32) bool {
        return qtc.QStringListModel_MoveRows(@ptrCast(self), @ptrCast(sourceParent), sourceRow, count, @ptrCast(destinationParent), destinationChild);
    }

    /// Allows for overriding the related default method
    pub fn OnMoveRows(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, i32, i32, ?*anyopaque, i32) callconv(.c) bool) void {
        qtc.QStringListModel_OnMoveRows(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMoveRows(self: ?*anyopaque, sourceParent: ?*anyopaque, sourceRow: i32, count: i32, destinationParent: ?*anyopaque, destinationChild: i32) bool {
        return qtc.QStringListModel_QBaseMoveRows(@ptrCast(self), @ptrCast(sourceParent), sourceRow, count, @ptrCast(destinationParent), destinationChild);
    }

    pub fn ItemData(self: ?*anyopaque, index: ?*anyopaque) QMap<int, QVariant> {
        return qtc.QStringListModel_ItemData(@ptrCast(self), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnItemData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) QMap<int, QVariant>) void {
        qtc.QStringListModel_OnItemData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseItemData(self: ?*anyopaque, index: ?*anyopaque) QMap<int, QVariant> {
        return qtc.QStringListModel_QBaseItemData(@ptrCast(self), @ptrCast(index));
    }

    pub fn SetItemData(self: ?*anyopaque, index: ?*anyopaque, roles: *const QMap<int, QVariant>) bool {
        return qtc.QStringListModel_SetItemData(@ptrCast(self), @ptrCast(index), @ptrCast(roles));
    }

    /// Allows for overriding the related default method
    pub fn OnSetItemData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, *const QMap<int, QVariant>) callconv(.c) bool) void {
        qtc.QStringListModel_OnSetItemData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetItemData(self: ?*anyopaque, index: ?*anyopaque, roles: *const QMap<int, QVariant>) bool {
        return qtc.QStringListModel_QBaseSetItemData(@ptrCast(self), @ptrCast(index), @ptrCast(roles));
    }

    pub fn Sort(self: ?*anyopaque, column: i32, order: i32) void {
        qtc.QStringListModel_Sort(@ptrCast(self), column, @intCast(order));
    }

    /// Allows for overriding the related default method
    pub fn OnSort(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, i32) callconv(.c) void) void {
        qtc.QStringListModel_OnSort(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSort(self: ?*anyopaque, column: i32, order: i32) void {
        qtc.QStringListModel_QBaseSort(@ptrCast(self), column, @intCast(order));
    }

    pub fn StringList(self: ?*anyopaque, ) []const u8 {
        return qtc.QStringListModel_StringList(@ptrCast(self));
    }

    pub fn SetStringList(self: ?*anyopaque, strings: []const u8) void {
        qtc.QStringListModel_SetStringList(@ptrCast(self), @ptrCast(strings));
    }

    pub fn SupportedDropActions(self: ?*anyopaque, ) i32 {
        return qtc.QStringListModel_SupportedDropActions(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnSupportedDropActions(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QStringListModel_OnSupportedDropActions(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSupportedDropActions(self: ?*anyopaque, ) i32 {
        return qtc.QStringListModel_QBaseSupportedDropActions(@ptrCast(self));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStringListModel_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstringlistmodel.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStringListModel_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstringlistmodel.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QStringListModel_Delete(@ptrCast(self));
    }
};

File: binding/src/libqstringlistmodel.cpp
#include <QAbstractListModel>
#include <QModelIndex>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QStringListModel>
#include <QVariant>
#include <qstringlistmodel.h>
#include "libqstringlistmodel.h"
#include "libqstringlistmodel.hxx"

QStringListModel* QStringListModel_new() {
	 return new VirtualQStringListModel();
}

QStringListModel* QStringListModel_new2(const libqt_list strings) {
	 return new VirtualQStringListModel(QList<QString>());
}

QStringListModel* QStringListModel_new3(QObject* parent) {
	 return new VirtualQStringListModel(parent);
}

QStringListModel* QStringListModel_new4(const libqt_list strings, QObject* parent) {
	 return new VirtualQStringListModel(QList<QString>(), parent);
}

libqt_string QStringListModel_Tr(const char* s) {
	QString _ret = QStringListModel::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QStringListModel_RowCount(const QStringListModel* self, const QModelIndex* parent) {
	return self->rowCount(*parent);
}

QModelIndex* QStringListModel_Sibling(const QStringListModel* self, int row, int column, const QModelIndex* idx) {
	return new QModelIndex(self->sibling(row, column, *idx));
}

QVariant* QStringListModel_Data(const QStringListModel* self, const QModelIndex* index, int role) {
	return new QVariant(self->data(*index, role));
}

bool QStringListModel_SetData(QStringListModel* self, const QModelIndex* index, const QVariant* value, int role) {
	return self->setData(*index, *value, role);
}

bool QStringListModel_ClearItemData(QStringListModel* self, const QModelIndex* index) {
	return self->clearItemData(*index);
}

int QStringListModel_Flags(const QStringListModel* self, const QModelIndex* index) {
	return self->flags(*index);
}

bool QStringListModel_InsertRows(QStringListModel* self, int row, int count, const QModelIndex* parent) {
	return self->insertRows(row, count, *parent);
}

bool QStringListModel_RemoveRows(QStringListModel* self, int row, int count, const QModelIndex* parent) {
	return self->removeRows(row, count, *parent);
}

bool QStringListModel_MoveRows(QStringListModel* self, const QModelIndex* sourceParent, int sourceRow, int count, const QModelIndex* destinationParent, int destinationChild) {
	return self->moveRows(*sourceParent, sourceRow, count, *destinationParent, destinationChild);
}

libqt_map QStringListModel_ItemData(const QStringListModel* self, const QModelIndex* index) {
	auto _ret = self->itemData(*index);
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}

bool QStringListModel_SetItemData(QStringListModel* self, const QModelIndex* index, const libqt_map roles) {
	return self->setItemData(*index, QMap<int, QVariant>());
}

void QStringListModel_Sort(QStringListModel* self, int column, int order) {
	self->sort(column, static_cast<Qt::SortOrder>(order));
}

libqt_list QStringListModel_StringList(const QStringListModel* self) {
	auto _ret = self->stringList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QStringListModel_SetStringList(QStringListModel* self, const libqt_list strings) {
	self->setStringList(QList<QString>());
}

int QStringListModel_SupportedDropActions(const QStringListModel* self) {
	return self->supportedDropActions();
}

libqt_string QStringListModel_Tr2(const char* s, const char* c) {
	QString _ret = QStringListModel::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QStringListModel_Tr3(const char* s, const char* c, int n) {
	QString _ret = QStringListModel::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QStringListModel_QBaseRowCount(const QStringListModel* self, const QModelIndex* parent) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_RowCount_IsBase(true);
	return vqstringlistmodel->rowCount(*parent);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnRowCount(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_RowCount_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_RowCount_Callback>(slot));
}
}

// Base class handler implementation
QModelIndex* QStringListModel_QBaseSibling(const QStringListModel* self, int row, int column, const QModelIndex* idx) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Sibling_IsBase(true);
	return new QModelIndex(vqstringlistmodel->sibling(row, column, *idx));
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnSibling(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Sibling_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_Sibling_Callback>(slot));
}
}

// Base class handler implementation
QVariant* QStringListModel_QBaseData(const QStringListModel* self, const QModelIndex* index, int role) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Data_IsBase(true);
	return new QVariant(vqstringlistmodel->data(*index, role));
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnData(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Data_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_Data_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseSetData(QStringListModel* self, const QModelIndex* index, const QVariant* value, int role) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SetData_IsBase(true);
	return vqstringlistmodel->setData(*index, *value, role);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnSetData(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SetData_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_SetData_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseClearItemData(QStringListModel* self, const QModelIndex* index) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_ClearItemData_IsBase(true);
	return vqstringlistmodel->clearItemData(*index);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnClearItemData(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_ClearItemData_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_ClearItemData_Callback>(slot));
}
}

// Base class handler implementation
int QStringListModel_QBaseFlags(const QStringListModel* self, const QModelIndex* index) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Flags_IsBase(true);
	return vqstringlistmodel->flags(*index);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnFlags(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Flags_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_Flags_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseInsertRows(QStringListModel* self, int row, int count, const QModelIndex* parent) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_InsertRows_IsBase(true);
	return vqstringlistmodel->insertRows(row, count, *parent);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnInsertRows(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_InsertRows_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_InsertRows_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseRemoveRows(QStringListModel* self, int row, int count, const QModelIndex* parent) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_RemoveRows_IsBase(true);
	return vqstringlistmodel->removeRows(row, count, *parent);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnRemoveRows(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_RemoveRows_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_RemoveRows_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseMoveRows(QStringListModel* self, const QModelIndex* sourceParent, int sourceRow, int count, const QModelIndex* destinationParent, int destinationChild) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_MoveRows_IsBase(true);
	return vqstringlistmodel->moveRows(*sourceParent, sourceRow, count, *destinationParent, destinationChild);
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnMoveRows(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_MoveRows_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_MoveRows_Callback>(slot));
}
}

// Base class handler implementation
libqt_map QStringListModel_QBaseItemData(const QStringListModel* self, const QModelIndex* index) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_ItemData_IsBase(true);
	auto _ret = vqstringlistmodel->itemData(*index);
	libqt_map _map;
	_map.len = _ret.size();
	_map.keys = nullptr;
	_map.values = nullptr;
	return _map;
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnItemData(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_ItemData_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_ItemData_Callback>(slot));
}
}

// Base class handler implementation
bool QStringListModel_QBaseSetItemData(QStringListModel* self, const QModelIndex* index, const libqt_map roles) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SetItemData_IsBase(true);
	return vqstringlistmodel->setItemData(*index, QMap<int, QVariant>());
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnSetItemData(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SetItemData_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_SetItemData_Callback>(slot));
}
}

// Base class handler implementation
void QStringListModel_QBaseSort(QStringListModel* self, int column, int order) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Sort_IsBase(true);
	vqstringlistmodel->sort(column, static_cast<Qt::SortOrder>(order));
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnSort(QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_Sort_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_Sort_Callback>(slot));
}
}

// Base class handler implementation
int QStringListModel_QBaseSupportedDropActions(const QStringListModel* self) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SupportedDropActions_IsBase(true);
	return vqstringlistmodel->supportedDropActions();
}
}

// Auxiliary method to allow providing re-implementation
void QStringListModel_OnSupportedDropActions(const QStringListModel* self, intptr_t slot) {
	auto* vqstringlistmodel = dynamic_cast<const VirtualQStringListModel*>(self);
	if (vqstringlistmodel && vqstringlistmodel->isVirtualQStringListModel) {
vqstringlistmodel->setQStringListModel_SupportedDropActions_Callback(reinterpret_cast<VirtualQStringListModel::QStringListModel_SupportedDropActions_Callback>(slot));
}
}

void QStringListModel_Delete(QStringListModel* self) {
    delete self;
}

